### PR TITLE
Converts puppet forge module entries into puppetfile git entries

### PR DESCRIPTION
**Pull request by logicminds**
* This feature is useful when you want to bring all the forge modules
    in house on the local git server.  This will auto lookup
    the module's homepage url and use that as the git source.

    Sometimes the author doesn't provide the homepage url so this
    will fail and the code isn't always tagged to the version so
    some manual intervention will be required before final approval
    of conversion output.

Example: `bundle exec rake r10k:print_git_conversion`

Converts the following

```ruby
# # In addition to the component modules listed here, the 'site' directory
# # includes 'role' and 'profile' modules. The 'role' module contains
# # Puppet classes that constitute a machine role or business function.
# #
# # Start Forge Modules
# # puppetlabs/windows pack
mod 'puppetlabs/reboot'
mod 'puppetlabs/registry',                 '1.1.4'
mod 'puppetlabs/powershell',               '2.1.0'
mod 'puppetlabs/acl',                      '1.1.2'
mod 'puppetlabs/wsus_client',              '1.0.3'
mod 'puppetlabs/chocolatey',               '3.0.0'
mod 'puppet/download_file',                '2.1.0'
mod 'puppet/windowsfeature',               '2.1.0'
mod 'puppet/windows_env',                  '2.3.0'
mod 'noma4i/windows_updates',              '0.2.2'
mod 'ianoberst/xml_fragment',              '1.0.2'
```
into

```ruby
mod 'reboot',
  :git => 'https://github.com/puppetlabs/puppetlabs-reboot',
  :ref => 'v4.1.0'

mod 'registry',
  :git => 'https://github.com/puppetlabs/puppetlabs-registry',
  :ref => '1.1.4'

mod 'powershell',
  :git => 'https://github.com/puppetlabs/puppetlabs-powershell',
  :ref => '2.1.0'

mod 'acl',
  :git => 'https://github.com/puppetlabs/puppetlabs-acl',
  :ref => '1.1.2'

mod 'wsus_client',
  :git => 'https://github.com/puppetlabs/puppetlabs-wsus_client',
  :ref => '1.0.3'

mod 'chocolatey',
  :git => 'https://github.com/puppetlabs/puppetlabs-chocolatey',
  :ref => '3.0.0'

mod 'download_file',
  :git => 'https://github.com/voxpupuli/puppet-download_file',
  :ref => 'v2.1.0'

mod 'windowsfeature',
  :git => 'https://github.com/voxpupuli/puppet-windowsfeature',
  :ref => 'v2.1.0'

mod 'windows_env',
  :git => 'https://github.com/voxpupuli/puppet-windows_env',
  :ref => 'v2.3.0'

mod 'windows_updates',
  :git => 'https://github.com/noma4i/puppet-windows_updates',
  :ref => 'bad url or tag 0.2.2 is missing'

mod 'xml_fragment',
  :git => 'https://github.com/Areson/xml_fragment',
  :ref => 'bad url or tag 1.0.2 is missing'
```

### CODE REVIEW

**Comment by logicminds** [comment]
I need to beef up the tests but was looking for feedback with the output.

This doesn't replace the puppetfile and only provides content to be put into a puppetfile.  Which reminds me I need to add some docs around this.

Does the task name look good?

File: lib/ra10ke/dependencies.rb [hunk 1]
```diff
@@ -138,6 +138,43 @@ def print_table(mods)
...
+
+      PuppetForge.user_agent = "ra10ke/#{Ra10ke::VERSION}"
+      puppetfile = get_puppetfile
+      puppetfile.load!
+      PuppetForge.host = puppetfile.forge if puppetfile.forge =~ /^http/
+
+      # ignore file allows for "don't tell me about this"
+      ignore_modules = []
+      if File.exist?('.r10kignore')
+        ignore_modules = File.readlines('.r10kignore').each {|l| l.chomp!}
```
Comment (bastelfreak) [inline lib/ra10ke/dependencies.rb:156]: that file would be a list of modules that should not be converted? Is that file already used somewhere else? Otherwise it should be documented.

File: lib/ra10ke/dependencies.rb [hunk 2]
```diff
@@ -138,6 +138,43 @@ def print_table(mods)
...
+      end
+      forge_mods = puppetfile.modules.find_all { |mod| mod.instance_of?(R10K::Module::Forge) && mod.v3_module.homepage_url? }
```
Comment (bastelfreak) [inline lib/ra10ke/dependencies.rb:158]: What is `homepage_url`? the `project_page` attribute from metadata.json or `source`?

File: lib/ra10ke/dependencies.rb (hunk 1 above)
Comment (logicminds) [reply lib/ra10ke/dependencies.rb:156]: Yeah I just copied it from the task below this one.  I assumed it was a ra10ke feature.  Not something I added but just continued to support.

**Review by bastelfreak — COMMENTED** [review]
I like the idea and left some minor inline comments. based on the task name I'm not sure if I expect some output or a rewritten Puppetfile. Which would you prefer? I think the task name should reflect that (the description already does). Or maybe offer two tasks, one for rewriting, one for printing?

File: lib/ra10ke/dependencies.rb (hunk 2 above)
Comment (logicminds) [reply lib/ra10ke/dependencies.rb:158]: So I think this is were some contention might be because homepage_url is a forge api thing and source is a metadata.json thing.  Both can be right and wrong and might even conflict.  Not sure which one to use so I just picked the one from the API.  Even so, it is still confusing and probably not the best.

I should get the metadata source url and compare.

**Comment by bastelfreak** [comment]
https://forgeapi.puppet.com/#tag/Module-Operations/operation/getModule isn't really helpful :D

**Comment by logicminds** [comment]
> I like the idea and left some minor inline comments. based on the task name I'm not sure if I expect some output or a rewritten Puppetfile. Which would you prefer? I think the task name should reflect that (the description already does). Or maybe offer two tasks, one for rewriting, one for printing?

I have a follow up task to this one that replaces the git urls with internal urls and namespaces.  I'll think about the task name a bit.  Maybe something with the word print in it.

print_git_format?

**Comment by bastelfreak** [comment]
https://forgeapi.puppet.com/v3/modules/puppet-borg
I think `homepage_url` maps to `project_page` from metadata.json. Instead of using `homepage_url` you should probably use `metadata.source`?

**Comment by bastelfreak** [comment]
> print_git_format

I think that's a suitable name :+1:

File: lib/ra10ke/dependencies.rb (hunk 1 above)
Comment (bastelfreak) [reply lib/ra10ke/dependencies.rb:156]: Interesting, never saw it previously :D

**Comment by logicminds** [comment]
@bastelfreak I am happy with these changes and my additional test found some edge cases I needed to fix.  Ready to merge if you approve.

**Review by bastelfreak — APPROVED** [review]

File: lib/ra10ke/dependencies.rb [hunk 3]
```diff
@@ -138,6 +138,49 @@ def print_table(mods)
...
+     
+      threads = forge_mods.map do |mod|
+       Thread.new do 
+          source_url = mod.v3_module.attributes.dig(:current_release, :metadata, :source) || mod.v3_module.homepage_url 
+          # git:// does not work with ls-remote command, convert to https
+          source_url = source_url.gsub('git://', 'https://')
```
Comment (bastelfreak) [inline lib/ra10ke/dependencies.rb:164]: IMO metadata-lint should only allow https links for the git repos
